File: src/models/obligation.ts
export class Obligation implements ObligationInfoResponse {
    public id: string;
    public obligationType: number;
    public counterpartyId: string;
    public cfoId: string;
    public amount: number;
    public currency: string;
    public dueDate: number;
    public status: number;
    public paidAmount: number;
    public comment: string;

    private constructor(id: string, obligationType: number, counterpartyId: string, cfoId: string, amount: number, currency: string, dueDate: number, status: number, paidAmount: number, comment: string) {
        this.id = id;
        this.obligationType = obligationType;
        this.counterpartyId = counterpartyId;
        this.cfoId = cfoId;
        this.amount = amount;
        this.currency = currency;
        this.dueDate = dueDate;
        this.status = status;
        this.paidAmount = paidAmount;
        this.comment = comment;
    }

    public toCreateRequest(): ObligationCreateRequest {
        return {
            obligationType: this.obligationType,
            counterpartyId: this.counterpartyId,
            cfoId: this.cfoId,
            amount: this.amount,
            currency: this.currency,
            dueDate: this.dueDate,
            status: this.status,
            paidAmount: this.paidAmount,
            comment: this.comment
        };
    }

    public toModifyRequest(): ObligationModifyRequest {
        return {
            id: this.id,
            obligationType: this.obligationType,
            counterpartyId: this.counterpartyId,
            cfoId: this.cfoId,
            amount: this.amount,
            currency: this.currency,
            dueDate: this.dueDate,
            status: this.status,
            paidAmount: this.paidAmount,
            comment: this.comment
        };
    }

    public clone(): Obligation {
        return new Obligation(this.id, this.obligationType, this.counterpartyId, this.cfoId, this.amount, this.currency, this.dueDate, this.status, this.paidAmount, this.comment);
    }

    public static of(response: ObligationInfoResponse): Obligation {
        return new Obligation(response.id, response.obligationType, response.counterpartyId, response.cfoId, response.amount, response.currency, response.dueDate, response.status, response.paidAmount, response.comment);
    }

    public static ofMulti(responses: ObligationInfoResponse[]): Obligation[] {
        const obligations: Obligation[] = [];
        for (const response of responses) {
            obligations.push(Obligation.of(response));
        }
        return obligations;
    }

    public static createNew(obligationType: number = 1): Obligation {
        return new Obligation('', obligationType, '0', '0', 0, 'RUB', 0, 1, 0, '');
    }
}

export interface ObligationCreateRequest {
    readonly obligationType: number;
    readonly counterpartyId: string;
    readonly cfoId: string;
    readonly amount: number;
    readonly currency: string;
    readonly dueDate: number;
    readonly status: number;
    readonly paidAmount: number;
    readonly comment: string;
}

export interface ObligationModifyRequest {
    readonly id: string;
    readonly obligationType: number;
    readonly counterpartyId: string;
    readonly cfoId: string;
    readonly amount: number;
    readonly currency: string;
    readonly dueDate: number;
    readonly status: number;
    readonly paidAmount: number;
    readonly comment: string;
}

export interface ObligationDeleteRequest {
    readonly id: string;
}

export interface ObligationInfoResponse {
    readonly id: string;
    readonly obligationType: number;
    readonly counterpartyId: string;
    readonly cfoId: string;
    readonly amount: number;
    readonly currency: string;
    readonly dueDate: number;
    readonly status: number;
    readonly paidAmount: number;
    readonly comment: string;
}

File: pkg/models/obligation.go
package models

// ObligationType represents obligation type
type ObligationType byte

const (
	OBLIGATION_TYPE_RECEIVABLE ObligationType = 1
	OBLIGATION_TYPE_PAYABLE    ObligationType = 2
)

// ObligationStatus represents obligation status
type ObligationStatus byte

const (
	OBLIGATION_STATUS_ACTIVE   ObligationStatus = 1
	OBLIGATION_STATUS_PARTIAL  ObligationStatus = 2
	OBLIGATION_STATUS_PAID     ObligationStatus = 3
)

// Obligation represents obligation data stored in database
type Obligation struct {
	ObligationId    int64            `xorm:"PK"`
	Uid             int64            `xorm:"INDEX(IDX_obligation_uid_deleted) NOT NULL"`
	Deleted         bool             `xorm:"INDEX(IDX_obligation_uid_deleted) NOT NULL"`
	ObligationType  ObligationType   `xorm:"NOT NULL DEFAULT 1"`
	CounterpartyId  int64            `xorm:"NOT NULL DEFAULT 0"`
	CfoId           int64            `xorm:"NOT NULL DEFAULT 0"`
	Amount          int64            `xorm:"NOT NULL DEFAULT 0"`
	Currency        string           `xorm:"VARCHAR(3) NOT NULL DEFAULT 'RUB'"`
	DueDate         int64            `xorm:"NOT NULL DEFAULT 0"`
	Status          ObligationStatus `xorm:"NOT NULL DEFAULT 1"`
	PaidAmount      int64            `xorm:"NOT NULL DEFAULT 0"`
	Comment         string           `xorm:"VARCHAR(255) NOT NULL DEFAULT ''"`
	CreatedUnixTime int64
	UpdatedUnixTime int64
	DeletedUnixTime int64
}

type ObligationGetRequest struct {
	Id int64 `form:"id,string" binding:"required,min=1"`
}

type ObligationCreateRequest struct {
	ObligationType ObligationType   `json:"obligationType"`
	CounterpartyId int64            `json:"counterpartyId,string"`
	CfoId          int64            `json:"cfoId,string"`
	Amount         int64            `json:"amount"`
	Currency       string           `json:"currency" binding:"required,max=3"`
	DueDate        int64            `json:"dueDate"`
	Status         ObligationStatus `json:"status"`
	PaidAmount     int64            `json:"paidAmount"`
	Comment        string           `json:"comment" binding:"max=255"`
}

type ObligationModifyRequest struct {
	Id             int64            `json:"id,string" binding:"required,min=1"`
	ObligationType ObligationType   `json:"obligationType"`
	CounterpartyId int64            `json:"counterpartyId,string"`
	CfoId          int64            `json:"cfoId,string"`
	Amount         int64            `json:"amount"`
	Currency       string           `json:"currency" binding:"required,max=3"`
	DueDate        int64            `json:"dueDate"`
	Status         ObligationStatus `json:"status"`
	PaidAmount     int64            `json:"paidAmount"`
	Comment        string           `json:"comment" binding:"max=255"`
}

type ObligationDeleteRequest struct {
	Id int64 `json:"id,string" binding:"required,min=1"`
}

type ObligationInfoResponse struct {
	Id             int64            `json:"id,string"`
	ObligationType ObligationType   `json:"obligationType"`
	CounterpartyId int64            `json:"counterpartyId,string"`
	CfoId          int64            `json:"cfoId,string"`
	Amount         int64            `json:"amount"`
	Currency       string           `json:"currency"`
	DueDate        int64            `json:"dueDate"`
	Status         ObligationStatus `json:"status"`
	PaidAmount     int64            `json:"paidAmount"`
	Comment        string           `json:"comment"`
}

func (o *Obligation) ToObligationInfoResponse() *ObligationInfoResponse {
	return &ObligationInfoResponse{
		Id:             o.ObligationId,
		ObligationType: o.ObligationType,
		CounterpartyId: o.CounterpartyId,
		CfoId:          o.CfoId,
		Amount:         o.Amount,
		Currency:       o.Currency,
		DueDate:        o.DueDate,
		Status:         o.Status,
		PaidAmount:     o.PaidAmount,
		Comment:        o.Comment,
	}
}
